dictionary.lisp

```diff
@@ -0,0 +1,120 @@
+#|
+ This file is a part of Qtools-UI
+ (c) 2019 Shirakumo http://tymoon.eu (shinmera@tymoon.eu)
+ Author: Michał "phoe" Herda <phoe@disroot.org>
+|#
+
+(in-package #:org.shirakumo.qtools.ui)
+(in-readtable :qtools)
+
+(define-widget dictionary (qwidget)
+  ((empty-browser-text :accessor empty-browser-text
+                       :initarg :empty-browser-text)
+   (not-found-text :accessor not-found-text
+                   :initarg :not-found-text))
+  (:default-initargs
+    :empty-browser-text
+    "<i><p align=center>Type your query below and hit Search.</p></i>"
+    :not-found-text
+    "<i><p align=center>The entry for \"~A\" was not found.</p><i>"))
+
+(define-subwidget (dictionary layout) (q+:make-qgridlayout)
+  (setf (q+:layout dictionary) layout
+        (q+:contents-margins layout) (values 0 0 0 0)))
+
+(define-subwidget (dictionary browser)
+    (make-instance 'browser :dictionary dictionary)
+  (q+:add-widget layout browser 0 0 1 2)
+  (setf (q+:html browser) empty-browser-text))
+
+(define-subwidget (dictionary input) (q+:make-qlineedit)
+  (q+:add-widget layout input 1 0))
+
+(defun make-text-qtoolbutton (text)
+  (let ((button (q+:make-qtoolbutton)))
+    (setf (q+:text button) text
+          (q+:tool-button-style button)
+          (q+:qt.tool-button-text-only))
+    button))
+
+(define-subwidget (dictionary button) (make-text-qtoolbutton "Search")
+  (q+:add-widget layout button 1 1)
+  (setf (q+:focus input) (q+:qt.other-focus-reason)))
+
+(defun trim-whitespace (string)
+  (string-trim '(#\Space #\Newline #\Tab) string))
+
+(define-slot (dictionary search) ()
+  (declare (connected button (clicked)))
+  (declare (connected input (return-pressed)))
+  (let ((text (trim-whitespace (q+:text input))))
+    (when (string/= "" text)
+      (let ((response (htmlize-wordnet (wordnet:wordnet-describe* text))))
+        (if (string/= "" response)
+            (setf (q+:html browser) response)
+            (setf (q+:html browser)
+                  (format NIL not-found-text text))))))) ;; TODO parametrize text
+
+(define-slot (dictionary set-focus-from-browser set-focus-from-browser)
+             ((old "QWidget*") (new "QWidget*"))
+  (declare (connected qt:*qapplication* (focus-changed "QWidget*" "QWidget*")))
+  (%set-focus-from-browser dictionary new))
+
+(defmethod set-focus-from-browser ((dictionary dictionary) old (new qobject))
+  (%set-focus-from-browser dictionary new))
+
+(defmethod set-focus-from-browser ((dictionary dictionary) old new))
+
+(defun %set-focus-from-browser (dictionary new)
```

Comment:
What's the point of this extra function? Why not just use the method?

Reply:
Fixed.